package/upgrade/lib.sh

```diff
@@ -580,4 +580,79 @@ EOF
 
   # wait status only when enabled and already AddonDeploySuccessful
   wait_for_addon_upgrade_deployment $name $namespace $enabled $curstatus
-}
\ No newline at end of file
+}
+
+patch_grafana_dashboard_harvester_configmap() {
+  local cm=$1
+  local jsonname=$2
+  local originValuesfile=$3
+  local newValuesfile=$4
+  local changed=0
+
+  # if there are differences, diff returns 1, else 0
+  diff $originValuesfile $newValuesfile || changed=$?
+  if [[ $changed == 1 ]]; then
+    # add 4 spaces to each line
+    sed -i -e 's/^/    /' $newValuesfile
+    local newvalues=$(<$newValuesfile)
+    rm -f $newValuesfile
+    local patchfile="/tmp/configmappatch.yaml"
+
+cat > $patchfile <<EOF
+data:
+  $jsonname: |
+$newvalues
+EOF
+
+    kubectl patch configmap $cm -n cattle-dashboards --patch-file $patchfile --type merge || echo "patch configmap $cm failed"
+    rm -f $patchfile
+  else
+    # the configmap may have been updated or patched
+    echo "need not patch configmap $cm, it is up-to-date"
+  fi
+
+  rm -f $originValuesfile
+  rm -f $newValuesfile
+}
+
+patch_grafana_dashboard_harvester_vm_detail() {
+  local EXIT_CODE=0
+  local cm=harvester-vm-detail-dashboard
+  local originValuesfile="/tmp/a1.yaml"
+  local newValuesfile="/tmp/a2.yaml"
+  rm -f $originValuesfile
+  echo "patch_grafana_dashboard_harvester_vm_detail"
+  kubectl get configmap -n cattle-dashboards $cm -ojsonpath="{.data['harvester_vm_info_details\.json']}" > $originValuesfile || EXIT_CODE=$?
+  if [[ $EXIT_CODE -gt 0 ]]; then
+    echo "fail to get configmap $cm, do not try to patch it"
+    return 0
+  fi
+
+  cp $originValuesfile $newValuesfile
+  sed -i -e 's/kubevirt_vmi_vcpu_seconds{/kubevirt_vmi_vcpu_seconds_total{/g' $newValuesfile
```

Comment:
Can we use `jq` to manipulate the JSON?

Reply:
@bk201 @Yu-Jack:

The reason why `jq` was not adopted in this patch:

Both `harvester_vm_dashboard.json` and `harvester_vm_info_details.json` in those 2 configmaps are using nested map & list to configure the data, with roughly 800 lines and 700 lines of text.

When replacing `kubevirt_vmi_vcpu_seconds` in `harvester_vm_dashboard.json`, the JSON path is like below, there are 2 list-index. It will be either fragile when simply hard-code indexing them (e.g. user added/removed some customized panel in the configmap and the index is not as assumed), or certainly complex to locate the target object by keyword searching &/ matching.


JSON path of `kubevirt_vmi_vcpu_seconds` in one occurance:
```
.["Panels"][0]["targets"][0]["expr"] = "$newvalue"
```

On the other hand, `sed` has it's advantage of processing the complex text file. The code can correctly patch the target or skip the target when it has already been patched.